### PR TITLE
feat(tracing): support psycopg3

### PR DESCRIPTION
**Pull request by wconti27**
This PR adds support for the newly released [Psycopg3](https://www.psycopg.org/psycopg3/docs/index.html) package. 

Changes made:
- change existing psycopg2 package housed within `ddtrace.contrib.psycopg` to `ddtrace.contrib.psycopg2`
- use existing psycopg directory for psycopg3 patch (package name of `psycopg` on pypy
- Abstracted most cursor / connection tracing to within the psycopg3 package (`ddtrace.contrib.psycopg`). 
- Added support for Async Cursor and Connection classes
- Added support for shortcuts: new constructor `psycopg.Cursor(connection)` and class method `connection.execute`
- Added tests for new functionality

Extension patching? Json Patching? Adapter patching. Haven't touched any other non-cursor or non-connection modules.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Author is aware of the performance implications of this PR as reported in the benchmarks PR comment.

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer is aware of, and discussed the performance implications of this PR as reported in the benchmarks PR comment.

### CODE REVIEW

**Comment by ZStriker19** [comment]
@wconti27  could you add a release note for this? `reno new <name_of_release_note>` example: https://github.com/DataDog/dd-trace-py/blob/1.x/releasenotes/notes/add-graphql-core-support-fd20dcbebfcaa8c7.yaml

File: ddtrace/contrib/django/patch.py [hunk 1]
```diff
@@ -77,20 +77,20 @@
 
 
 _NotSet = object()
-psycopg_cursor_cls = Psycopg2TracedCursor = _NotSet
+psycopg_cursor_cls = PsycopgTracedCursor = _NotSet
 
 
 def patch_conn(django, conn):
```
Comment (ZStriker19) [inline ddtrace/contrib/django/patch.py:83]: should we add additional testing for this, to make sure that we pick out we instrument correctly whether its psycopg or pscyopg2?

File: ddtrace/contrib/utils.py
[the remark this answers is not in the document]
Comment (wconti27) [reply ddtrace/contrib/utils.py:1]: I guess `psycopg/utils.py` works. Does anyone else from Python team have input?

File: riotfile.py [hunk 2]
```diff
@@ -1143,6 +1143,24 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
                 ),
             ],
         ),
+        Venv(
+            name="psycopg",
+            command="pytest {cmdargs} tests/contrib/psycopg",
+            pkgs={"psycopg2-binary": latest, "pytest-asyncio": latest},
```
Comment (ZStriker19) [inline riotfile.py:1149]: why psycopg2 here?

File: riotfile.py [hunk 3]
```diff
@@ -1143,6 +1143,24 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
...
+            venvs=[
+                # Venv(
```
Comment (ZStriker19) [inline riotfile.py:1151]: remove commented code?

File: ddtrace/contrib/utils.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+
+def _unpatch_extensions(_extensions):
+    # we must patch extensions all the time (it's pretty harmless) so split
+    # from global patching of connections. must be idempotent.
+    for original, module, func, _ in _extensions:
+        setattr(module, func, original)
+
+
+#
```
Comment (ZStriker19) [inline ddtrace/contrib/utils.py:107]: delete the comment lines?

File: ddtrace/contrib/utils.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    # we must patch extensions all the time (it's pretty harmless) so split
+    # from global patching of connections. must be idempotent.
+    for _, module, func, wrapper in _extensions:
+        if not hasattr(module, func) or isinstance(getattr(module, func), wrapt.ObjectProxy):
+            continue
+        wrapt.wrap_function_wrapper(module, func, wrapper)
+
+
+def _unpatch_extensions(_extensions):
+    # we must patch extensions all the time (it's pretty harmless) so split
```
Comment (ZStriker19) [inline ddtrace/contrib/utils.py:101]: Nit: I like comments, but do we want the same one that we have above in `_patch_extensions`?

File: tests/contrib/django/test_django_snapshots.py [hunk 6]
```diff
@@ -125,8 +125,8 @@ def test_404_exceptions(client):
 def psycopg2_patched(transactional_db):
     from django.db import connections
 
-    from ddtrace.contrib.psycopg.patch import patch
-    from ddtrace.contrib.psycopg.patch import unpatch
+    from ddtrace.contrib.psycopg2.patch import patch
```
Comment (ZStriker19) [inline tests/contrib/django/test_django_snapshots.py:128]: can Django use psycopg3 instead? If so should we test that as well?

File: tests/contrib/psycopg/test_psycopg.py [hunk 7]
```diff
@@ -44,7 +38,8 @@ def tearDown(self):
         unpatch()
 
     def _get_conn(self, service=None):
-        conn = psycopg2.connect(**POSTGRES_CONFIG)
+        print(POSTGRES_CONFIG)
```
Comment (ZStriker19) [inline tests/contrib/psycopg/test_psycopg.py:41]: remove this?

File: tests/contrib/psycopg/test_psycopg.py [hunk 8]
```diff
@@ -181,7 +176,7 @@ def test_opentracing_propagation(self):
             )
             assert_is_measured(self.get_spans()[1])
 
-    @skipIf(PSYCOPG2_VERSION < (2, 5), "context manager not available in psycopg2==2.4")
+    # @skipIf(psycopg_VERSION < (2, 5), "context manager not available in psycopg==2.4")
```
Comment (ZStriker19) [inline tests/contrib/psycopg/test_psycopg.py:179]: remove comment?

File: tests/contrib/psycopg/test_psycopg_snapshot.py [hunk 9]
```diff
@@ -13,45 +14,69 @@
...
     """
 
     env = os.environ.copy()
     env["DD_PSYCOPG_TRACE_CONNECT"] = "true"
     out, err, status, pid = run_python_code_in_subprocess(code, env=env)
+    print(err)
     assert status == 0, err
+    print(status)
+    print(out)
+    # assert (
```
Comment (ZStriker19) [inline tests/contrib/psycopg/test_psycopg_snapshot.py:77]: do we want to remove this comment, or uncomment it?

File: tests/contrib/psycopg/test_psycopg.py [hunk 10]
```diff
@@ -308,16 +260,15 @@ def test_composed_query_identifier(self):
                 dict(name="postgres.query", resource=query.as_string(db)),
             )
 
-    @snapshot()
-    @skipIf(PSYCOPG2_VERSION < (2, 7), "SQL string composition not available in psycopg2<2.7")
+    @snapshot(async_mode=True)
     def test_composed_query_encoding(self):
```
Comment (ZStriker19) [inline tests/contrib/psycopg/test_psycopg.py:264]: Does the fact that there's no new snapshot test for this mean that we get the same snapshot for psycopg2 and psycopg3? Just want to make sure we ran it and confirmed that.

**Review by ZStriker19 — APPROVED** [review]

**Pull request merged**